compose/project.py

```diff
@@ -176,20 +184,23 @@ def get_links(self, service_dict):
     def get_volumes_from(self, service_dict):
         volumes_from = []
         if 'volumes_from' in service_dict:
-            for volume_name in service_dict.get('volumes_from', []):
+            for volume_from_config in service_dict.get('volumes_from', []):
+                volume_from_spec = parse_volume_from_spec(volume_from_config)
+                # Get service
                 try:
-                    service = self.get_service(volume_name)
-                    volumes_from.append(service)
+                    service_name = self.get_service(volume_from_spec.source)
+                    volume_from_spec = VolumeFromSpec(service_name, volume_from_spec.mode)
                 except NoSuchService:
                     try:
-                        container = Container.from_id(self.client, volume_name)
-                        volumes_from.append(container)
+                        container_name = Container.from_id(self.client, volume_from_spec.source)
+                        volume_from_spec = VolumeFromSpec(container_name, volume_from_spec.mode)
                     except APIError:
                         raise ConfigurationError(
                             'Service "%s" mounts volumes from "%s", which is '
                             'not the name of a service or container.' % (
                                 service_dict['name'],
-                                volume_name))
+                                volume_from_spec.source))
+                volumes_from.append(volume_from_spec)
```

Comment:
Yeh I agree with you it is a bit strange. I did have a go at refactoring it, but threw away my attempt as it wasn't right(but useful as I subsequently learnt more about `volumes_from`). If we have time during bug fix we can take another crack at refactoring it but agree it's not blocking for rn.